cmd/job_run_list.go

```diff
@@ -0,0 +1,156 @@
+	cmd.Flags().StringVar(&host, "host", defaultHost, "Optimus service endpoint url")
+	cmd.Flags().StringVar(&startDate, "start_date", "", "start date of job run")
+	cmd.Flags().StringVar(&endDate, "end_date", "", "end date of job run")
+	cmd.RunE = func(c *cli.Command, args []string) error {
+		jobName := args[0]
+		l.Info(fmt.Sprintf("Requesting status for project %s, job %s from %s",
+			projectName, jobName, host))
+		arg := jobRunCmdArg{
+			projectName: projectName,
+			jobName:     jobName,
+			startDate:   startDate,
+			endDate:     endDate,
+		}
+		return getJobRunList(l, host, arg)
+	}
+	return cmd
+}
+
+func getJobRunList(l log.Logger, host string, arg jobRunCmdArg) error {
+	// get the last schedule run
+	if arg.startDate == "" && arg.endDate == "" {
+		return callJobRun(l, host, arg, getJobRunRequest)
+	}
+	// get runs between start_date and end_date (both start_date and end_date are inclusive)
+	return callJobRun(l, host, arg, getJobRunRequestWithDate)
+}
+
+func callJobRun(l log.Logger, host string, arg jobRunCmdArg, fn getRequestFn) error {
+	var err error
+	dialTimeoutCtx, dialCancel := context.WithTimeout(context.Background(), OptimusDialTimeout)
+	defer dialCancel()
+
+	var conn *grpc.ClientConn
+	if conn, err = createConnection(dialTimeoutCtx, host); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			l.Info("can't reach optimus service, timing out")
+		}
+		return err
+	}
+	defer conn.Close()
+
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), jobStatusTimeout)
+	defer cancel()
+
+	run := pb.NewJobRunServiceClient(conn)
+	req, err := fn(arg)
+	if err != nil {
+		return err
+	}
+
+	spinner := NewProgressBar()
+	spinner.Start("please wait...")
+	jobRunResponse, err := run.JobRun(timeoutCtx, req)
+	spinner.Stop()
+	if err != nil {
+		return fmt.Errorf("request failed for job %s: %w", arg.jobName, err)
+	}
+
+	jobRuns := jobRunResponse.GetJobRuns()
+
+	for _, jobRun := range jobRuns {
+		l.Info(fmt.Sprintf("%s - %s", jobRun.GetScheduledAt().AsTime(), jobRun.GetState()))
+	}
+	l.Info(coloredSuccess("\nFound %d jobRun instances.", len(jobRuns)))
+	return err
+}
+
+func getJobRunRequest(arg jobRunCmdArg) (*pb.JobRunRequest, error) {
+	req := &pb.JobRunRequest{
+		ProjectName: arg.projectName,
+		JobName:     arg.jobName,
+	}
+	return req, nil
+}
+
+func getJobRunRequestWithDate(arg jobRunCmdArg) (*pb.JobRunRequest, error) {
+	var req *pb.JobRunRequest
+	dateRange, err := validateDateRange(arg.startDate, arg.endDate)
+	if err != nil {
+		return req, fmt.Errorf("request failed for job %s: %w", arg.jobName, err)
+	}
+	req = &pb.JobRunRequest{
+		ProjectName: arg.projectName,
+		JobName:     arg.jobName,
+		StartDate:   timestamppb.New(dateRange.start),
+		EndDate:     timestamppb.New(dateRange.end),
+	}
+	return req, nil
+}
+
+func validateDateRange(startDate, endDate string) (dateRange, error) {
```

Comment:
validate shouldn't return anything other than error

Reply:
addressed as part of latest commit